flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/DefaultCompletedCheckpointStore.java

```diff
@@ -118,33 +112,22 @@ public boolean requiresExternalizedCheckpoints() {
 	}
 
 	/**
-	 * Gets the latest checkpoint from ZooKeeper and removes all others.
-	 *
-	 * <p><strong>Important</strong>: Even if there are more than one checkpoint in ZooKeeper,
-	 * this will only recover the latest and discard the others. Otherwise, there is no guarantee
-	 * that the history of checkpoints is consistent.
+	 * Recover all the valid checkpoints from state handle store. All the successfully recovered checkpoints will
+	 * be added to {@link #completedCheckpoints} sorted by checkpoint id.
 	 */
 	@Override
 	public void recover() throws Exception {
-		LOG.info("Recovering checkpoints from ZooKeeper.");
+		LOG.info("Recovering checkpoints from {}.", checkpointStateHandleStore);
 
 		// Get all there is first
-		List<Tuple2<RetrievableStateHandle<CompletedCheckpoint>, String>> initialCheckpoints;
-		while (true) {
-			try {
-				initialCheckpoints = checkpointsInZooKeeper.getAll();
-				break;
-			}
-			catch (ConcurrentModificationException e) {
-				LOG.warn("Concurrent modification while reading from ZooKeeper. Retrying.");
-			}
-		}
+		final List<Tuple2<RetrievableStateHandle<CompletedCheckpoint>, String>> initialCheckpoints =
+			checkpointStateHandleStore.getAll();
```

Comment:
1. IIUC, we will never throw a `ConcurrentModificationException` in `StateHandleStore#getAllAndLock`. So this retry does not take any effect.
2. For `ZooKeeperStateHandleStore` implementation, it could retry internally to avoid concurrent modifications.
3. For `KubernetesStateHandleStore` implementation, we do not need retry since we could get all the data in one time.

Reply:
Just want to double check why we needed to catch `ConcurrentModificationException` at the first place, so we can be sure that removing the loop won't cause any problem.

Reply:
After checking more previous commits, it seems that the retry loop in `ZooKeeperStateHandleStore#recover` is there from the first commit. Also in the `ZooKeeperStateHandleStore`, I do not notice that `getAllAndLock` or `getAllSortedByNameAndLock`(has replaced by the former) could not throw a `ConcurrentModificationException` in all the history.

@tillrohrmann Could you share some insights on this?

Reply:
I cannot explain why there is the `ConcurrentModificationException`. To me it looks that it is not really required.